api/dbadapters/presto.ts

```diff
@@ -0,0 +1,228 @@
+  public async tables(): Promise<dataform.ITarget[]> {
+    let tables: dataform.ITarget[] = [];
+    await Promise.all(
+      (await this.schemas()).map(async schemaString => {
+        const [catalog, schema] = schemaString.split(".");
+        const result = await prestoExecute(this.client, `show tables from ${catalog}.${schema}`);
+        tables = tables.concat(
+          flatten(result.data).map(table => ({ database: catalog, schema, table }))
+        );
+      })
+    );
+    return tables;
+  }
+
+  public async table(target: dataform.ITarget): Promise<dataform.ITableMetadata> {
+    let columnsResult: IPrestoExecutionResult;
+    try {
+      columnsResult = await prestoExecute(this.client, `describe ${resolveTarget(target)}`);
+    } catch (e) {
+      // This probably failed because the table doesn't exist, so ignore as the information isn't necessary.
+    }
+    // TODO: Add primitives mapping.
+    // Columns are structured [column name, type (primitive), extra, comment]. This can be
+    // seen under columnsResult.columns; instead they could be dynamically populated using this info.
+    const fields =
+      columnsResult?.data.map(column =>
+        dataform.Field.create({ name: column[0], description: column[3] })
+      ) || [];
+    return dataform.TableMetadata.create({
+      // TODO: Add missing table metadata items.
+      target,
+      fields
+    });
+  }
+
+  public async preview(target: dataform.ITarget, limitRows: number = 10): Promise<any[]> {
+    const { rows } = await this.execute(
+      `select * from ${target.database}.${target.schema}.${target.name} limit ${limitRows}`
+    );
+    return rows;
+  }
+
+  public async persistedStateMetadata(): Promise<dataform.IPersistedTableMetadata[]> {
+    // TODO.
+    return [];
+  }
+
+  public async persistStateMetadata() {
+    // TODO.
+  }
+
+  public async setMetadata(action: dataform.IExecutionAction): Promise<void> {
+    // TODO.
+  }
+
+  public async close() {
+    // Not required, as data is transferred using the Presto REST api.
+  }
+}
+
+export function prestoExecute(
+  client: Presto.Client,
+  statement?: string
+): Promise<IPrestoExecutionResult> {
+  return new Promise((resolve, reject) => {
+    const result: IPrestoExecutionResult = {};
+    client.execute({
+      query: statement,
+      cancel: () => {
+        return false;
+      },
+      state: (error: any, queryId: string, stats: Presto.IPrestoClientStats) => {
+        result.error = error;
+        result.queryId = queryId;
+        result.stats = stats;
+      },
+      columns: (error: any, columns: Presto.IPrestoClientColumnMetaData[]) => {
+        result.error = error;
+        result.columns = columns;
+      },
+      data: (
+        error: any,
+        data: Presto.PrestoClientColumnDatum[],
+        columns: Presto.IPrestoClientColumnMetaData[],
+        stats: Presto.IPrestoClientStats
+      ) => {
+        result.error = error;
+        result.data = data;
```

Comment:
this isn't right, the API docs say that `data` may be called >1 times so this just overwrites previous data fetches.

Reply:
Have updated it to `concat` data instead. I'm wondering of columns does the same thing. Having a full project for the integration test should catch  any weird behaviour here.